### ISSUE TITLE
Downloading Deleted Files... only to delete them?

### ISSUE THREAD

**Issue by michaelkrieger** [icloud-photos-downloader/icloud_photos_downloader#511]
## Overview
With the latest update I've noticed a large number of deleted images- many more than I would ever delete each day (60-80).

Looking at the logs, the app is downloading images, and in the same run deleting the very same files it just downloaded (by file name).
The logs have nothing interesting in the preamble or file list, except, after it reviews the deduplicated files and downloaded files, it reviews the deleted files and deletes a bunch of files it just downloaded (about 60-80 each time).

Why is it downloading the recently deleted images and subsequently deleting them?

## Steps to Reproduce
I add 30-40 photos to my phone
icloudPD downloads 122
icloudPD deletes 62
The next day it downloads the same deleted images and re-deletes them.

## Expected Behavior
It should not download recently deleted.

## Actual Behavior
It downloads recently deleted and then deletes them.

**Comment by AndreyNikiforov** [comment]
@michaelkrieger Thanks for sharing details. IIUC delete functionality runs at the end of the processing if enabled and this behavior was there from the initial implementation.

> With the latest update...

What update are you referring to? Latest official release is 1.7.2 from Jan 2021...

**Comment by michaelkrieger** [comment]
> @michaelkrieger Thanks for sharing details. IIUC delete functionality runs at the end of the processing if enabled and this behavior was there from the initial implementation.
> What update are you referring to? Latest official release is 1.7.2 from Jan 2021...
Yes using icloudpd, version 1.7.2, but using it within the docker-icloudpd container which adds lots of wrapper scripts.
I'm guessing the bug is better addressed on that side.  Will update here if it turns out to be something with icloudpd.  There were some recent updates to the scripts in that container.

**Comment by AndreyNikiforov** [comment]
Related to #196

**Comment by holomekc** [comment]
Hmm I think this still happens right? Running the tool via docker twice, both runs downloads for example IMG_1.HEIC, but then in the deletion steps realizes that it should be removed. 

In my head it should be possible to check the recently deleted "folder" to keep in memory, which images, videos are already deleted and during download phase they should not be downloaded but instead skipped.

Or am I mistaken?

**Comment by AndreyNikiforov** [comment]
> Hi @AndreyNikiforov Hmm I think this still happens right? Running the tool via docker twice, both runs downloads for example IMG_1.HEIC, but then in the deletion step realize that it should be removed.
> 
> In my head it should be possible to check the recently deleted "folder" to keep in memory, which images, videos are already deleted and during download phase they should not be downloaded but instead skipped.
> 
> Or am I mistaken?

Haven't not look at the yet ;-( My gut feeling is that [name-id7](https://icloud-photos-downloader.github.io/icloud_photos_downloader/naming.html#duplicates) policy may have that problem to a lesser degree (or not at all)

**Comment by The-Croz** [comment]
I am still experiencing this same issue. Is there any fix for this? icloudpd running in Docker downloads and deletes the same photos everyday.

**Comment by boredazfcuk** [comment]
You have two files with the same name. One in your photo stream and one in 'Recently Deleted'. Remove the one in 'Recently Deleted' manually, or wait 30 days for it to expire and be removed automatically.

**Comment by The-Croz** [comment]
@boredazfcuk gotcha. Will this continue to happen as I delete more photos in iCloud going forward? Or will it stop once this initial file conflict is resolved?

**Comment by boredazfcuk** [comment]
Depends on how the duplicated files are being created. I found that saving files out of the Telegram app saved them with the same name, so the files in the photo stream were actually different to the ones in the 'Recently Deleted' album.

I also found I could replicate the problem by putting the phone in low-power mode (or just by having a low battery), taking a photo, deleting it, then taking another... this would result in it re-using the name of the file that had just been removed.

After that, switch off low power mode (or charge the battery), and it will sync the device to iCloud. Now you should see the problem.

I get around it by just not deleting photos immediately. I'll either make sure my library is fully sync'd first, but more often than not, I'll take the lazy approach and just delete stuff that's more than a day old.

**Comment by julme** [comment]
We hit the same failure mode in a real external-library workflow: iCloudPD downloaded a photo during a successful active-library sync, then `--auto-delete` moved the local file to its trash because an item from Recently Deleted reconstructed to the same local path. Immich subsequently marked the external asset offline and moved it to its own trash.

The underlying problem appears to be that auto-delete derives local paths from mutable/path-level attributes (creation date, filename, folder structure, rendition/version), while the iCloud API provides a stable asset ID. The delete decision is therefore not tied to the identity of the file that was actually downloaded. A distinct or stale API record resolving to the same path can remove a valid current asset.

Could auto-delete persist an atomic mapping from `PhotoAsset.id` to the exact local paths actually materialized during a successful full active-library scan? Then, for each Recently Deleted asset:

- resolve only paths recorded for that same stable asset ID;
- refuse to delete when the manifest is absent, incomplete, stale, or ambiguous;
- refuse destructive actions unless the current run has a complete active-library inventory and the asset ID is absent from it;
- retain the current move-to-trash behavior, but log the asset ID, manifest generation, and exact paths used.

This would make auto-delete fail closed instead of deleting based on a reconstructed path alone. It would also correctly handle collision suffixes, renamed renditions, and future filename/date changes.

I can help test or submit a patch if this design is acceptable.
